### PR TITLE
Slim dependencies

### LINKED ISSUE #2391: Rethinking Zarr's core dependencies
I'd like to open the conversation about what Zarr's core dependencies are for 3.0. Currently, this looks like: 

https://github.com/zarr-developers/zarr-python/blob/11312534ebe683d73cbbcc2da9e88933cb00cc14/pyproject.toml#L25-L34

Some of these are not used anymore (`asciitree` and `fasteners`) so those can safely go.

Then there is `fsspec` and `crc32c`. These are only needed for the `RemoteStore` and `ShardingCodec`, respectively. What do we think about making these optional?

One proposed diff in our dependencies would look something like:

```diff
 dependencies = [
-    'asciitree',
     'numpy>=1.25',
-    'fasteners',
-    'numcodecs>=0.10.2',
-    'fsspec>2024',
-    'crc32c',
+    'numcodecs>=0.12',
     'typing_extensions',
     'donfig',
 ]

 [project.optional-dependencies]
+remote = [
+    "fsspec",
+]
+sharding = [
+    "crc32c",
+]
```

Notes:
- fsspec is pure python with no dependencies so is not a particular heavy dependency.
- crc32c could potentially move into numcodecs, right?

### PR DESCRIPTION
**Pull request by jhamman**
This PR implements the main ideas from #2391, removing `crc32c` and `fsspec` as core dependencies and adding them to the `sharding` and `remote` groups respectively. Imports are adjusted in minor ways and we now take advantage of `pytest.importorskip` in a few new places.

TODO: if we decide to go this route, we should have a separate part of the test matrix that includes optional dependencies.

Closes #2391 
Fixes https://github.com/zarr-developers/zarr-python/issues/735

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Review by dstansby — CHANGES_REQUESTED** [review]
I think we're coming to a consensus in https://github.com/zarr-developers/zarr-python/issues/2391 (?) of
- Move dependency of `crc32c` to numcodecs (https://github.com/zarr-developers/numcodecs/issues/610)
- Make `fsspec` optional

For the second one, would be good to add some docs explaining how to install the `remote` extra, and what functionality it enables.

File: pyproject.toml [hunk 1]
```diff
@@ -54,17 +51,19 @@ license = {text = "MIT License"}
 keywords = ["Python", "compressed", "ndimensional-arrays", "zarr"]
 
 [project.optional-dependencies]
+remote = [
```
Comment (jhamman) [inline pyproject.toml:54]:
```suggestion
fsspec = [
```

**Comment by jhamman** [comment]
Something I learned over the weekend is that `crc32c` is LGPL Licensed. Regardless of whether we get `crc32c` into `numcodecs`, we may want to pick a different implementation ([googleapis/python-crc32c](https://github.com/googleapis/python-crc32c) is Apache 2.0 licenced).

File: pyproject.toml [hunk 2]
```diff
@@ -26,13 +26,10 @@ maintainers = [
...
-    'asciitree',
     'numpy>=1.25',
-    'numcodecs>=0.10.2',
-    'fsspec>2024',
-    'crc32c',
-    'typing_extensions',
-    'donfig',
+    'numcodecs>=0.12',
+    'typing_extensions>=4.9',
+    'donfig>=0.8.1',
```
Comment (jhamman) [inline pyproject.toml:32]:
```suggestion
    'donfig>=0.8.1',
    'crc32c',
```

File: pyproject.toml [hunk 3]
```diff
@@ -54,17 +51,19 @@ license = {text = "MIT License"}
 keywords = ["Python", "compressed", "ndimensional-arrays", "zarr"]
 
 [project.optional-dependencies]
+fsspec = [
+    "fsspec>=2023.10.0",
+]
+sharding = [
+    "crc32c>=2.4",
+]
```
Comment (jhamman) [inline pyproject.toml:59]:
```suggestion
```

**Pull request closed without merging**